### PR TITLE
Added CIDR to whitelist (IP and CIDRs are allowed). Bug with config file opening.

### PR DESCRIPTION
**Pull request by marcinguy**
Added CIDR to whitelist (IP and CIDRs are allowed). Bug with file opening, the file sometimes does not open (!). Weird NGINX behavior.

### CODE REVIEW

File: naxsi_src/naxsi_runtime.c [hunk 1]
```diff
@@ -401,6 +401,58 @@ int nx_find_pass_in_hash(ngx_http_request_t *req,
...
+int nx_find_pass_in_array(ngx_http_request_t *req,
+                                           ngx_str_t *mstr,
+                                           ngx_http_dummy_loc_conf_t *cf,
+                                           enum DUMMY_MATCH_ZONE zone) {
+  uint32_t netip;
+  uint32_t netmask;
+  uint i = 0;
+  for (i = 0; i < cf->passr_headers_array->nelts; i++) {
+   
+    if(cidr_to_ip_and_mask(((char *)((ngx_str_t *)cf->passr_headers_array->elts)[i].data),&netip, &netmask) < 0) {
```
Comment (wargio) [inline naxsi_src/naxsi_runtime.c:430]: something is missing in this `if` block

**Review by wargio — CHANGES_REQUESTED** [review]

File: naxsi_src/naxsi_runtime.c [hunk 2]
```diff
@@ -401,6 +401,58 @@ int nx_find_pass_in_hash(ngx_http_request_t *req,
...
+    }   
+    u_char a, b, c, d;
+    
+    char *ipstr= (char *) mstr->data; // value to check
+    uint32_t ip = 0;
+    if (sscanf(ipstr, "%hhu.%hhu.%hhu.%hhu", &a, &b, &c, &d) < 4) {
```
Comment (wargio) [inline naxsi_src/naxsi_runtime.c:436]: same here

**Comment by wargio** [comment]
this is great! thanks for these awesome contribs! :)

File: naxsi_src/naxsi_utils.c [hunk 3]
```diff
@@ -653,25 +653,21 @@ ngx_http_wlr_finalize_hashtables(ngx_conf_t *cf, ngx_http_dummy_loc_conf_t  *dlc
     }
   }
 
-  FILE *file;
-  char fname[1024];
+  FILE *file=NULL;
+  char *fname=NULL;
```
Comment (wargio) [inline naxsi_src/naxsi_utils.c:657]: spaces.

File: naxsi_src/naxsi_utils.c [hunk 4]
```diff
@@ -653,25 +653,21 @@ ngx_http_wlr_finalize_hashtables(ngx_conf_t *cf, ngx_http_dummy_loc_conf_t  *dlc
...
   if(dlc->whitelist_file)
   {
-    size_t fname_size = dlc->whitelist_file->len;
-    if (fname_size >= sizeof(fname)) {
-      fname_size = sizeof(fname);
-    }
-    memcpy(fname, dlc->whitelist_file->data, fname_size);
+    fname=malloc(strlen((char *)dlc->whitelist_file->data)*sizeof(char)+1);
```
Comment (wargio) [inline naxsi_src/naxsi_utils.c:660]: i think a `fname = calloc(1024, sizeof(char))` is enough and you need to check also that `fname` is a valid pointer (not NULL).

File: naxsi_src/naxsi_utils.c [hunk 5]
```diff
@@ -713,53 +712,79 @@ ngx_http_wlr_finalize_hashtables(ngx_conf_t *cf, ngx_http_dummy_loc_conf_t  *dlc
...
+      if(!strstr(&descs[i],"/"))
       {
-        for (i = 0; i < nlines; i++) {
-          arr_node = (ngx_hash_key_t *)ngx_array_push(elements);
-          arr_node->key = (names[i]);
-          arr_node->key_hash =  ngx_hash_key_lc(arr_node->key.data, arr_node->key.len);
-          arr_node->value = &descs[i];
-        }
+        ips_cnt++;
+        ips[j++]=descs[i];
```
Comment (wargio) [inline naxsi_src/naxsi_utils.c:750]: spaces `ips[j++] = descs[i];`.

File: naxsi_src/naxsi_utils.c [hunk 6]
```diff
@@ -713,53 +712,79 @@ ngx_http_wlr_finalize_hashtables(ngx_conf_t *cf, ngx_http_dummy_loc_conf_t  *dlc
...
+      }else{
```
Comment (wargio) [inline naxsi_src/naxsi_utils.c:751]: spaces `} else {`

**Review by wargio — CHANGES_REQUESTED** [review]

File: naxsi_src/naxsi_runtime.c (hunk 1 above)
Comment (marcinguy) [reply naxsi_src/naxsi_runtime.c:429]: Corrected

File: naxsi_src/naxsi_utils.c (hunk 3 above)
Comment (marcinguy) [reply naxsi_src/naxsi_utils.c]: Corrected

File: naxsi_src/naxsi_utils.c (hunk 5 above)
Comment (marcinguy) [reply naxsi_src/naxsi_utils.c]: Corrected

File: naxsi_src/naxsi_utils.c (hunk 6 above)
Comment (marcinguy) [reply naxsi_src/naxsi_utils.c]: Corrected

File: naxsi_src/naxsi_utils.c (hunk 4 above)
Comment (marcinguy) [reply naxsi_src/naxsi_utils.c]: Corrected

**Comment by marcinguy** [comment]
> this is great! thanks for these awesome contribs! :)

Pleasure. Cool project :)

File: naxsi_src/naxsi_utils.c [hunk 7]
```diff
@@ -653,25 +653,24 @@ ngx_http_wlr_finalize_hashtables(ngx_conf_t *cf, ngx_http_dummy_loc_conf_t  *dlc
...
-  char fname[1024];
+  FILE *file = NULL;
+  char *fname = NULL;
   if(dlc->whitelist_file)
   {
-    size_t fname_size = dlc->whitelist_file->len;
-    if (fname_size >= sizeof(fname)) {
-      fname_size = sizeof(fname);
+    fname = calloc(1024, sizeof(char));
+    if(fname !== NULL)
```
Comment (wargio) [inline naxsi_src/naxsi_utils.c]: invalid check. `!=` is correct.

File: naxsi_src/naxsi_utils.c [hunk 8]
```diff
@@ -653,25 +653,24 @@ ngx_http_wlr_finalize_hashtables(ngx_conf_t *cf, ngx_http_dummy_loc_conf_t  *dlc
...
+    {
+      strncpy(fname, (char *)dlc->whitelist_file->data, strlen((char *)dlc->whitelist_file->data));
```
Comment (wargio) [inline naxsi_src/naxsi_utils.c:663]: you removed the check on strncpy like we had it before..

**Review by wargio — CHANGES_REQUESTED** [review]

File: naxsi_src/naxsi_utils.c (hunk 8 above)
Comment (marcinguy) [reply naxsi_src/naxsi_utils.c:663]: Don't know how and why but this:

```
  char fname[1024];
  if(dlc->whitelist_file)
  {
    size_t fname_size = dlc->whitelist_file->len;
    if (fname_size >= sizeof(fname)) {
      fname_size = sizeof(fname);
    }
    memcpy(fname, dlc->whitelist_file->data, fname_size);
```


Worked once in a while, sometimes, yes, sometimes no.

Comment (marcinguy) [reply naxsi_src/naxsi_utils.c:663]: Now it is hardcoded at 1024? What do you suggest? @wargio is it OK?

File: naxsi_src/naxsi_utils.c (hunk 7 above)
Comment (marcinguy) [reply naxsi_src/naxsi_utils.c]: Yep.

File: naxsi_src/naxsi_utils.c (hunk 8 above)
Comment (marcinguy) [reply naxsi_src/naxsi_utils.c:663]: The len was good, but the data was corrupted, unintilaized at the end of the string @wargio

Comment (marcinguy) [reply naxsi_src/naxsi_utils.c:663]: If you want to see it, you can try on your own ... maybe you will find the root cause @wargio

Comment (wargio) [reply naxsi_src/naxsi_utils.c:663]: just change that to `size_t fname_size = strlen((char *)dlc->whitelist_file->data);` and `if (fname_size >= 1024) {`

Comment (marcinguy) [reply naxsi_src/naxsi_utils.c:665]: Changed

File: naxsi_src/naxsi_utils.c [hunk 9]
```diff
@@ -653,25 +653,27 @@ ngx_http_wlr_finalize_hashtables(ngx_conf_t *cf, ngx_http_dummy_loc_conf_t  *dlc
...
   {
-    size_t fname_size = dlc->whitelist_file->len;
-    if (fname_size >= sizeof(fname)) {
-      fname_size = sizeof(fname);
+    fname = calloc(1024, sizeof(char));
+    if(fname != NULL)
+    {
+      size_t fname_size = strlen((char *)dlc->whitelist_file->data);
+      if(fname_size <= 1024) {
+      strncpy(fname, (char *)dlc->whitelist_file->data, strlen((char *)dlc->whitelist_file->data));
```
Comment (wargio) [inline naxsi_src/naxsi_utils.c:665]: add some indentation since it is inside the `if`

**Review by wargio — CHANGES_REQUESTED** [review]
besides the indentation issue, LGTM.

**Comment by wargio** [comment]
would be nice to add some tests.

File: naxsi_src/naxsi_utils.c (hunk 9 above)
Comment (marcinguy) [reply naxsi_src/naxsi_utils.c:665]: Corrected

**Comment by marcinguy** [comment]
> would be nice to add some tests.

I will work on this as time permits. However, this is not the highest prio for me now.  I know TDD :))))

Tested manually and it worked on my setup.

**Review by wargio — APPROVED** [review]

**Pull request merged**